src/plugins/bfetch/server/streaming/create_compressed_stream.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { promisify } from 'util';
+import { Observable } from 'rxjs';
+import { catchError, concatMap, finalize } from 'rxjs/operators';
+import { Logger } from 'src/core/server';
+import { Stream, PassThrough } from 'stream';
+import { constants, deflate } from 'zlib';
+
+const delimiter = '\n';
+
+async function zipMessageToStream(output: PassThrough, message: string) {
+  return new Promise(async (resolve, reject) => {
+    try {
+      const gzipped = await promisify(deflate)(message, {
+        flush: constants.Z_SYNC_FLUSH,
+      });
+      output.write(gzipped.toString('base64'));
+      output.write(delimiter);
+      resolve(undefined);
+    } catch (err) {
+      reject(err);
+    }
+  });
+}
+
+export const createCompressedStream = <Response>(
+  results: Observable<Response>,
+  logger: Logger
+): Stream => {
+  const output = new PassThrough();
+
+  const sub = results
+    .pipe(
+      concatMap((message: Response) => {
+        const strMessage = JSON.stringify(message);
+        return zipMessageToStream(output, strMessage);
```

Comment:
Curious why this isn't following the lower-level example provided here: https://github.com/elastic/kibana/pull/98658/
Didn't it work?

Reply:
Great question!
I did get the code to work using `zlib` streams, but the problem with it is that you can't use `base64` encoding.
This happens because base64 output is padded and hence `base64(a) + base64(b) != base64(a+b)`.
The result is that I either had to use `hex` encoding (which inflates the data much more than `base64`) or handle binary data (which is more complicated to parse on the client).
I thought that this version of the code is a good compromise, but I am open to continue the discussion. :-)

Reply:
I am very sorry but even after reading base64 https://github.com/elastic/kibana/pull/97994#issuecomment-832632293 explanation I think I am still missing some context and don't fully understand why do we need it and if we can work without it :(

Reply:
You can see [here](https://github.com/elastic/kibana/pull/97994/commits/a88372bd732a79120b0ab3d9652aa5727bd83fb7#diff-c1b2acef2b9531b73ed107b327247239c5c360893349ed0e26fae235cb8fbfdeR19) that I implemented the code like Mikhail suggested. 

It works, but it forces me to use `hex` encoding which [doubles the size](https://stackoverflow.com/a/40613599/372086) of the returned message.

### Why?

In the linked commit, the `data` callback gets called **multiple times** for a single message, which is the reason why I can't use `base64` on it. `base64` needs to be called on the final message, otherwise it would result in a corrupt encoding that can't be decoded by the client. 

This leaves us with the `binary` option. While it is indeed the most compact, it would force us to change the protocol from using `\n` separated messages (`\n` can't be used as a separator in binary data) to a more "classic" protocol, where we prefix each message with its length. 

It is possible of course, but to me, introducing the additional complexity of manipulating binary data in JS, for a ~3% size decrease feels like an overkill. Especially not in a bugfix PR, and considering we do want to explore a more long term solution with HTTP/2, Websockets or SSEs.

Reply:
I haven't looked at Mikhails example in details, but doesn't the browser handle all decompression in his example?

Reply:
@mshustov 

The message structure is 
`<gzipped message 1>\n<gzipped message 2>\n...` (The code for `\n` is `0a`) 

So if transferred as binary data, you'd have  `<some binary 1>0a<some binary 2>0a`.

If the chunks sent to the client are not aligned (and they are not on long messages), how would the client know that a given `0a` is a separator or just a part of the binary data.

If we use binary, I believe the format should be changed to `<message size><binary data1><message size><binary data 2>. And this is a bigger change that I wouldn't push as a bug fix.

Reply:
> If the chunks sent to the client are not aligned (and they are not on long messages), how would the client know that a given 0a is a separator or just a part of the binary data.

when client-side part decompresses a chunk, it receives `{ndjson}\n{ndjson}\n`. Can't we split the decompressed string by `\n`? Or you are worried about cases when `ndjson` contains `\n` separator, like:`{ data: 'aaaa \n bbbb' }`? How it's different from splitting by `\n` when decompression is disabled? https://github.com/elastic/kibana/blob/611df93efaf84e846cf594e62c3425d27868d449/src/plugins/bfetch/public/streaming/fetch_streaming.ts#L58

Reply:
The order is reverse :-)
The stream **first** needs to be split and only then **decompressed**.
The separator needs to separate compressed chunks of data, otherwise you won't be able to use chunks as the come in (which is the point of bfetch).

Reply:
Discussed offline. It's required for cases with a huge payload when the server might send compressed chunks in several batches. Since we don’t know the size of every batch, we have to use \n as a separator to signal to the client-side that the payload now can be decompressed.